python/gps/algorithm/algorithm_traj_opt.py

```diff
@@ -115,37 +115,35 @@ def _update_cost(self):
         ix = range(self.dX)
         iu = range(self.dX, self.dX + self.dU)
         init_samples = self.init_samples
-        demo_sum, samples_sum = {}, {}
-        # demoU = {i: self.demo_list.get_U() for i in xrange(self.M)}
-        # demoX = {i: self.demo_list.get_X() for i in xrange(self.M)}
-        demoX = self.demoX
-        demoU = self.demoU
-
         # itration_count + 1 distributions to evaluate
         # T: summed over time
-        samples_prob, demos_prob, samples_prob_T, demos_prob_T = {}, {}, {}, {}
+        samples_logprob, demos_logprob = {}, {}
         # number of demo distributions
-        Md = len(demoX)
+        Md = self._hyperparams['demo_Md']
+        demos_logiw, samples_logiw = {}, {}
+        demoU = {i: self.demo_list.get_U() for i in xrange(self.Md)}
+        demoX = {i: self.demo_list.get_X() for i in xrange(self.Md)}
+        demoO = {i: self.demo_list.get_obs() for i in xrange(self.Md)}
         demo_traj = {}
         # estimate demo distributions empirically
         for i in xrange(Md):
             if self._hyperparams['demo_distr_empest']:
                 demo_traj[i] = fit_emp_controllers(demoX[i], demoU[i])
         for i in xrange(M):
             # This code assumes a fixed number of samples per iteration/controller
-            samples_prob[i] = np.zeros((itr + Md, self.T, (self.N / M) * itr + init_samples))
-            demos_prob[i] = np.zeros((itr + Md, self.T, demoX[i].shape[0]))
+            samples_logprob[i] = np.zeros((itr + Md + 1, self.T, (self.N / M) * itr + init_samples))
+            demos_logprob[i] = np.zeros((itr + Md + 1, self.T, demoX[i].shape[0]))
             sample_i_X = self.cur[i].sample_list.get_X()
             sample_i_U = self.cur[i].sample_list.get_U()
```

Comment:
We're going to need to save all of the samples collected up until this point, not just the samples from the current iteration, similar to how all of the previous trajectories are stored.

Then, when we compute importance weights and update the cost, we do it for all of the existing samples. Note, that the rest of the learning code should still just use the samples from the current iteration (which are stored in self.cur)